### ISSUE TITLE
ssl.ca.location and related settings do not affect libcurl requests for OIDC token endpoints

### ISSUE THREAD

**Issue by lpsinger** [edenhill/librdkafka#3751]
libcurl is used to request the OIDC token endpoint, but libcurl's SSL settings are not configured by librdkafka.

Description
===========
The `ssl.ca.location` setting in particular is important because the system on which librdkafka is run may have CA certificates in a different location than the system on which librdkafka was built. However the SSL settings only affect the SSL context used for the broker connection, and do not affect the SSL context used by libcurl for HTTPS requests.

How to reproduce
================
N/A

Checklist
=========

**IMPORTANT**: We will close issues where the checklist has not been completed.

Please provide the following information:

 - [x] librdkafka version (release number or git tag): `<REPLACE with e.g., v0.10.5 or a git sha. NOT "latest" or "current">`
 - [ ] Apache Kafka version: `<REPLACE with e.g., 0.10.2.3>`
 - [ ] librdkafka client configuration: `<REPLACE with e.g., message.timeout.ms=123, auto.reset.offset=earliest, ..>`
 - [ ] Operating system: `<REPLACE with e.g., Centos 5 (x64)>`
 - [ ] Provide logs (with `debug=..` as necessary) from librdkafka
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
Right, I wonder though if they can always be considered to require the same CA chain, seeing how they could be run by different organizations.

Maybe we need a separate `sasl.oauthbearer.oidc.ssl.ca.location` property that defaults to `ssl.ca.location` ?

**Comment by lpsinger** [comment]
Indeed. I was hoping that there would be some guidance from the Kafka Java client config, but there is not.

But I think that the common case that both need to be able to find the OS CA bundle is the most important one.

**Comment by edenhill** [comment]
Okay, so I think we have three different scenarios we want to support:
 1. **inherit**: Use same CA chain as ssl.ca.location. This should be the default.
 2. **system**: Use the system default CA chain,.
 3. **/some/path**: Use a specific CA chain

**Comment by RafalSkolasinski** [comment]
I don't see any activity here so just wanted to check - is there currently any workaround for situation where CAs are not located under `/etc/ssl/certs/ca-certificates.crt`?

**Comment by lpsinger** [comment]
> I don't see any activity here so just wanted to check - is there currently any workaround for situation where CAs are not located under `/etc/ssl/certs/ca-certificates.crt`?

Our current workaround is to not use confluent-kafka-python's built-in KIP-768 support and instead to use a workalike oauth_cb written in pure Python. Here is an example in our project: https://github.com/nasa-gcn/gcn-kafka-python/blob/main/gcn_kafka/oidc.py

**Comment by RafalSkolasinski** [comment]
We are using golang clients in our context. If I understand you right @lpsinger, you are basically fetching JWT tokens on your own and pass it to Kafka client?

**Comment by lpsinger** [comment]
Yes. The oath callback mechanism exists in librdkafka too, so you could implement an analogous workaround in go.

**Comment by JohnPreston** [comment]
FYI I didn't face this issue with using Python images built based on Amazon Linux 2/2023, but indeed that's a problem it seems for all Debian-based builds. My workaround was to simply symlink the ca-bundle cert to the path the library complains it can't find it.

For example, see https://github.com/JohnPreston/kafka-overwatch/blob/main/Dockerfile#L21-L23

But that does warrant the idea of being able to provide the `sasl.oauth.ca.location` (or something along that name) as an option for the users to override the default.

**Comment by thomasnal** [comment]
@lpsinger This post is tremendously useful. I have implemented the callback, however, librdkafka seems to be bugged handling OAuth extensions together with the callback. Do you use extensions in your case?

The callback retrieves the token - thanks for the sample.  But Kafka fails the whole authentication with error,
```
SASL authentication error: Authentication failed: 1 extensions are invalid! 
They are: logicalCluster: CLUSTER_ID_MISSING_OR_EMPTY (after 5268ms in state AUTH_REQ) (_AUTHENTICATION)
```
I verified the extensions attribute is provided the correct value `'sasl.oauthbearer.extensions': 'logicalCluster=***,identityPoolId=***',`. Since I am connecting to confluent Kafka the extensions attribute is required.

**Comment by lpsinger** [comment]
> Do you use extensions in your case?

No, we are not setting `sasl.oauthbearer.extensions` in our project.

**Comment by martinfalke** [comment]
Found a simple fix [here](https://learn.microsoft.com/en-us/answers/questions/1605213/failed-to-retrieve-oidc-token-from-https-login-mic). They link to this issue and mentions @JohnPreston 's solution, but I managed to read past it when looking through this thread initially. Thanks for the solution! I'll post it explicitly here for reference, in case his repo link becomes unavailable/obsolete or someone reads as sloppily as I did:
```dockerfile
RUN mkdir -p /etc/pki/tls/certs/
# Link
RUN ln -s /usr/lib/ssl/certs/ca-certificates.crt /etc/pki/tls/certs/ca-bundle.crt
```
- Create the directory libcurl expects to find the bundle in
- Create a symbolic link where the bundle is expected to the "legacy" (SSL rather than TLS) path.

This worked for our AWS solution on an EC2 instance during initial testing of the broker connection (executed without 'RUN' in a sessions manager bash terminal), and by adding the commands to the Dockerfile of an ECS application in CI/CD pipelines for the real use case.
Do note that you may want to add the same check that John does to make sure you do not mess with existing bundles, as such:
```dockerfile
RUN if ! [ -d  /etc/pki/tls/certs/ ]; \
    then mkdir -p /etc/pki/tls/certs/; ln -s /usr/lib/ssl/cert.pem /etc/pki/tls/certs/ca-bundle.crt ;\
    fi
```

**Comment by adrian-gheorghe** [comment]
> Thanks. I have just been through Kafka source code. I figured out they expect the callback to return extensions as extra arguments in the tuple. For anyone else, it expects the extensions in form of a dict.
> 
> ext = {'logicalCluster':'***,'identityPoolId':'***'}
> principal = ''  # No complaints when empty string
> return token["access_token"], token["expires_at"], principal, ext

Thank you very much for this comment @thomasnal . It made me realize how to pass in these params in the dotnet client package 😁
